File: java/01209_remove_all_adjacent_duplicates_in_string_II.java
//       Java
// Time complexity - O (N)
// Space complexity - O (1)

// example: s = "deeedbbcccbdaa", k = 3
//          s = "deeedbbcccbdaa"

public class Solution {
    public String removeDuplicates(String s, int k) {
        Stack<Pair> stack = new Stack<>();

        for (char c : s.toCharArray()) {
            if (!stack.isEmpty() && stack.peek().character == c) {
                stack.peek().count++;
                if (stack.peek().count == k) {
                    stack.pop();
                }
            } else {
                stack.push(new Pair(c, 1));
            }
        }

        StringBuilder result = new StringBuilder();
        for (Pair p : stack) {
            result.append(String.valueOf(p.character).repeat(p.count));
        }
        return result.toString();
    }

    private static class Pair {
        char character;
        int count;

        Pair(char character, int count) {
            this.character = character;
            this.count = count;
        }
    }
}

File: python/01209_remove_all_adjacent_duplicates_in_string_II.py
#       Python
# Time complexity - O (N)
# Space complexity - O (1)

# example: s = "deeedbbcccbdaa", k = 3
#          s = "deeedbbcccbdaa"

class Solution:
    def removeDuplicates(self, s: str, k: int) -> str:
        stack = []  # character, count

        for char in s:
            if stack and stack[-1][0] == char:
                stack[-1][1] += 1
                if stack[-1][1] == k:
                    stack.pop()
            else:
                stack.append([char, 1])

        result = ''.join(char * count for char, count in stack)
        return result